# Python Task

> *We have data on page visits from tracking of one of the services over a certain period of time. During this time, the service was testing a new registration page. Can it be said that the new registration page is better than the old one?*
>
> *Determine which metrics can be used to assess the 'goodness' of the page and calculate them.*

In [1]:
import pandas as pd
import numpy as np

In [2]:
data = pd.read_csv('example_data')

In [3]:
data

user_id                      timestamp  \
0        148870bfa84777898359aaa8e120a373  2021-01-01 00:00:01.000000000   
1        ac3948ea43cb39cdc4e739004d252d0b  2021-01-01 00:00:01.445020335   
2        48a0df50d7ed1fcaaddf742b828b85e5  2021-01-01 00:00:10.566157670   
3        70fbdd335abb11a3d072b5de7b218048  2021-01-01 00:00:10.764937005   
4        48a0df50d7ed1fcaaddf742b828b85e5  2021-01-01 00:00:10.764937005   
...                                   ...                            ...   
1040166  a63250880822c619ecbbf9fa511d31cd  2021-04-07 13:20:16.775249492   
1040167  7fa83873bead5c5a52d6805570aba31d  2021-04-07 13:20:18.688737570   
1040168  1d716213d6f611f80592391ac61b5a5c  2021-04-07 13:20:18.688737570   
1040169  7fa83873bead5c5a52d6805570aba31d  2021-04-07 13:20:22.134689896   
1040170  c9be7da3b5c975cc2795f15d45f0390a  2021-04-07 13:20:34.394602584   

                event  
0             landing  
1             landing  
2               login  
3                main  
4               login  
...               ...  
1040166         login  
1040167  registration  
1040168          main  
1040169  registration  
1040170          main  

[1040171 rows x 3 columns]

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1040171 entries, 0 to 1040170
Data columns (total 3 columns):
 #   Column     Non-Null Count    Dtype 
---  ------     --------------    ----- 
 0   user_id    1040171 non-null  object
 1   timestamp  1040171 non-null  object
 2   event      1040171 non-null  object
dtypes: object(3)
memory usage: 23.8+ MB


In [5]:
data.describe()

user_id                      timestamp  \
count                            1040171                        1040171   
unique                            261130                         938229   
top     c9be7da3b5c975cc2795f15d45f0390a  2021-03-19 08:53:22.740489453   
freq                                5130                             12   

          event  
count   1040171  
unique        5  
top     landing  
freq     389350

In [6]:
data.nunique()

user_id      261130
timestamp    938229
event             5
dtype: int64

In [7]:
data['event'].unique()

array(['landing', 'login', 'main', 'registration', 'registration_new'],
      dtype=object)

In [8]:
data.groupby('event').nunique()

user_id  timestamp
event                               
landing            206953     355775
login               50485     244098
main                44635     327560
registration        24078      49812
registration_new     2540       5428

In [9]:
data['event'].value_counts()

landing             389350
main                336030
login               257227
registration         51943
registration_new      5621
Name: event, dtype: int64

In [10]:
data_ts = data
data_ts['timestamp'] = pd.to_datetime(data_ts['timestamp'])
data_ts.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1040171 entries, 0 to 1040170
Data columns (total 3 columns):
 #   Column     Non-Null Count    Dtype         
---  ------     --------------    -----         
 0   user_id    1040171 non-null  object        
 1   timestamp  1040171 non-null  datetime64[ns]
 2   event      1040171 non-null  object        
dtypes: datetime64[ns](1), object(2)
memory usage: 23.8+ MB


In [11]:
data_ts.head(20)

user_id                     timestamp  \
0   148870bfa84777898359aaa8e120a373 2021-01-01 00:00:01.000000000   
1   ac3948ea43cb39cdc4e739004d252d0b 2021-01-01 00:00:01.445020335   
2   48a0df50d7ed1fcaaddf742b828b85e5 2021-01-01 00:00:10.566157670   
3   70fbdd335abb11a3d072b5de7b218048 2021-01-01 00:00:10.764937005   
4   48a0df50d7ed1fcaaddf742b828b85e5 2021-01-01 00:00:10.764937005   
5   70fbdd335abb11a3d072b5de7b218048 2021-01-01 00:00:26.757563301   
6   48a0df50d7ed1fcaaddf742b828b85e5 2021-01-01 00:00:30.111507967   
7   7c4d355f03cfc54bd351b51fd6950bc8 2021-01-01 00:00:42.556088605   
8   0fd5fc803b51f27e8ab1dad44d12594e 2021-01-01 00:00:44.500571856   
9   6fe9147ebeb41c766a6f649239cce40b 2021-01-01 00:01:00.565173574   
10  3e2c632a66e8f884b30634327b43822a 2021-01-01 00:01:11.234373855   
11  4f2ba63e1426a17f3e089f30f3f3c737 2021-01-01 00:01:17.963347338   
12  ea07f8a72a7ad247e9f712de139f718d 2021-01-01 00:01:18.596746621   
13  3e2c632a66e8f884b30634327b43822a 2021-01-01 00:01:18.750451096   
14  3e2c632a66e8f884b30634327b43822a 2021-01-01 00:01:20.317246972   
15  9dd45f60cfad85074fb3a964683f3c25 2021-01-01 00:01:31.095477978   
16  3e2c632a66e8f884b30634327b43822a 2021-01-01 00:01:34.316237006   
17  3e2c632a66e8f884b30634327b43822a 2021-01-01 00:01:34.694580366   
18  0a36a903ff3440b4cf20322174490116 2021-01-01 00:01:41.759531120   
19  0a36a903ff3440b4cf20322174490116 2021-01-01 00:01:45.307747790   

           event  
0        landing  
1        landing  
2          login  
3           main  
4          login  
5           main  
6           main  
7        landing  
8           main  
9           main  
10       landing  
11          main  
12          main  
13       landing  
14       landing  
15  registration  
16       landing  
17       landing  
18         login  
19          main

In [12]:
sorted_data = data_ts.sort_values(by=['user_id', 'timestamp'])
sorted_data.head(50)


user_id                     timestamp  \
83898    00006e145e005308c5387dfbb3c9a490 2021-01-08 11:35:41.977195939   
536462   00008aab9f1597af45ad21aa141030aa 2021-02-17 00:36:00.319473015   
537547   00008aab9f1597af45ad21aa141030aa 2021-02-17 04:32:35.786731688   
537597   00008aab9f1597af45ad21aa141030aa 2021-02-17 04:42:10.180987082   
537624   00008aab9f1597af45ad21aa141030aa 2021-02-17 04:45:25.626196034   
537665   00008aab9f1597af45ad21aa141030aa 2021-02-17 04:50:38.128422496   
537700   00008aab9f1597af45ad21aa141030aa 2021-02-17 04:57:24.844030486   
537736   00008aab9f1597af45ad21aa141030aa 2021-02-17 05:03:17.211478741   
537757   00008aab9f1597af45ad21aa141030aa 2021-02-17 05:06:09.468095542   
537770   00008aab9f1597af45ad21aa141030aa 2021-02-17 05:08:58.010618286   
537771   00008aab9f1597af45ad21aa141030aa 2021-02-17 05:09:09.877435571   
236766   0000920f448655f105fa592fc22029c2 2021-01-21 21:05:40.841177844   
472741   000095d3c9d93db2db0327d7a1329e2d 2021-02-11 22:03:02.200545450   
472744   000095d3c9d93db2db0327d7a1329e2d 2021-02-11 22:03:22.471830090   
472751   000095d3c9d93db2db0327d7a1329e2d 2021-02-11 22:03:46.158900878   
473002   000095d3c9d93db2db0327d7a1329e2d 2021-02-11 22:47:26.658093831   
473003   000095d3c9d93db2db0327d7a1329e2d 2021-02-11 22:47:29.475738466   
558325   000095d3c9d93db2db0327d7a1329e2d 2021-02-19 09:21:22.376759940   
597957   000095d3c9d93db2db0327d7a1329e2d 2021-02-22 15:25:03.072415208   
868238   0000a840f208e41830a588db4745d128 2021-03-18 00:10:58.874420621   
868240   0000a840f208e41830a588db4745d128 2021-03-18 00:11:17.856938466   
852355   0000ef4ce7a22fdac5c651d180831096 2021-03-16 16:07:13.037395937   
388841   00011001c0146aa557eaf6865f7b0f97 2021-02-04 06:59:53.275869306   
388842   00011001c0146aa557eaf6865f7b0f97 2021-02-04 06:59:53.275869306   
618367   00015810e88ccc6ef242ce29e87d23fc 2021-02-24 04:59:33.905488613   
792147   0001a1a83bfe675a77dc1f4861b97db9 2021-03-10 20:04:03.244697000   
600946   0001bf9f9bd1c353baf64278f25541a0 2021-02-22 20:46:12.173034069   
211828   0001bfeb3f96ef37b1c1811afa14b1ce 2021-01-19 16:52:28.138519257   
817975   000212ce0574fc1389c571a91e56c05a 2021-03-13 14:42:28.475439475   
764906   0002c620bf43c7b0659be6f70b9d92eb 2021-03-08 15:21:10.826280080   
766144   0002c620bf43c7b0659be6f70b9d92eb 2021-03-08 17:13:29.584025808   
129208   00031d59f4d3dbb714a0bd768aa00944 2021-01-12 16:41:04.079152347   
129225   00031d59f4d3dbb714a0bd768aa00944 2021-01-12 16:45:23.638326457   
129228   00031d59f4d3dbb714a0bd768aa00944 2021-01-12 16:46:48.730481211   
129231   00031d59f4d3dbb714a0bd768aa00944 2021-01-12 16:47:05.955110914   
129251   00031d59f4d3dbb714a0bd768aa00944 2021-01-12 16:49:32.614467072   
129255   00031d59f4d3dbb714a0bd768aa00944 2021-01-12 16:50:10.110489634   
129364   00031d59f4d3dbb714a0bd768aa00944 2021-01-12 17:07:23.023772640   
129367   00031d59f4d3dbb714a0bd768aa00944 2021-01-12 17:07:53.074023906   
129431   00031d59f4d3dbb714a0bd768aa00944 2021-01-12 17:20:56.018160725   
129441   00031d59f4d3dbb714a0bd768aa00944 2021-01-12 17:23:27.350730691   
129447   00031d59f4d3dbb714a0bd768aa00944 2021-01-12 17:24:12.593399558   
129463   00031d59f4d3dbb714a0bd768aa00944 2021-01-12 17:27:01.379173607   
129464   00031d59f4d3dbb714a0bd768aa00944 2021-01-12 17:27:04.723702437   
129466   00031d59f4d3dbb714a0bd768aa00944 2021-01-12 17:27:36.475267751   
129467   00031d59f4d3dbb714a0bd768aa00944 2021-01-12 17:27:38.397978828   
129468   00031d59f4d3dbb714a0bd768aa00944 2021-01-12 17:28:05.572162358   
1033685  0003e7b564c8a85e7867eb23e34e956c 2021-04-06 23:34:46.954280044   
1033687  0003e7b564c8a85e7867eb23e34e956c 2021-04-06 23:34:48.946447037   
1033690  0003e7b564c8a85e7867eb23e34e956c 2021-04-06 23:34:56.321930867   

                event  
83898         landing  
536462          login  
537547   registration  
537597   registration  
537624           main  
537665          login  
537700           main  
537736          

## Research Design 

Unfortunately, we do not have exact metrics to test. Therefore, we will consider that there is a funnel:
**registration page &rarr; login**  

We will consider that a user who completes registration sooner or later will log in.  

Therefore, we will do the following for both 'registration' and 'registration_new' events separately:  

1. Count how many users visited a registration page AND later logged in. Technically, we will find amount of users who have both 'registration' and 'login' events and timestamp of the former one is less than the same of the latter one (i.e., **registration** was *before* **login**) 
<br>  

2. Count amount of users who have ever visited a registration page  
<br>  

3. Calculate proportion (ratio of 'register and log in' to 'register')
<br>  

4. Formulate and test hypotheses
<br>  

5. Make conclusions

In [38]:
# Count conversion for 'registration'
# Filter for 'registration' and 'login' events
registration_data = data[data['event'] == 'registration']
login_data = data[data['event'] == 'login']

# Merge the two DataFrames on 'user_id'
merged_data = registration_data.merge(login_data, on='user_id', suffixes=('_registration', '_login'))

# Group by 'user_id' and check if there's at least one case where 'login' timestamp is greater than 'registration' timestamp
user_has_condition = merged_data.groupby('user_id').apply(lambda x: (x['timestamp_login'] > x['timestamp_registration']).any())

# Count the number of users meeting the condition
count_reg_log = user_has_condition.sum()
distinct_registration_users = data[data['event'] == 'registration']['user_id'].nunique()
# count_reg = (data['event'] == 'registration').sum()
ratio_reg = count_reg_log / distinct_registration_users

# Print the count
print(f"Number of users with at least one case of 'login' having timestamp greater than 'registration': {count_reg_log}\nNumber of users with 'registration' values: {distinct_registration_users} \nRatio: {ratio_reg}")


Number of users with at least one case of 'login' having timestamp greater than 'registration': 8768
Number of users with 'registration' values: 24078 
Ratio: 0.3641498463327519


In [39]:
# Count conversion for 'registration_new'
# Filter for 'registration' and 'login' events
registration_data_new = data[data['event'] == 'registration_new']
login_data_new = data[data['event'] == 'login']

# Merge the two DataFrames on 'user_id'
merged_data_new = registration_data_new.merge(login_data_new, on='user_id', suffixes=('_registration_new', '_login'))

# Group by 'user_id' and check if there's at least one case where 'login' timestamp is greater than 'registration' timestamp
user_has_condition_new = merged_data_new.groupby('user_id').apply(lambda x: (x['timestamp_login'] > x['timestamp_registration_new']).any())

# Count the number of users meeting the condition
count_reg_log_new = user_has_condition_new.sum()
distinct_registration_users_new = data[data['event'] == 'registration_new']['user_id'].nunique()
# count_reg_new = (data['event'] == 'registration_new').sum()
ratio_reg_new = count_reg_log_new / distinct_registration_users_new

# Print the count
print(f"Number of users with at least one case of 'login' timestamp greater than 'registration_new': {count_reg_log_new}\nNumber of users with 'registration_new' values: {distinct_registration_users_new} \nRatio: {ratio_reg_new}")


Number of users with at least one case of 'login' timestamp greater than 'registration_new': 1007
Number of users with 'registration_new' values: 2540 
Ratio: 0.3964566929133858


In [40]:
# Check if there are any user_id's with both 'registration' and 'registration_new' events 
# Filter for 'registration' and 'registration_new' events
filtered_data = data[data['event'].isin(['registration', 'registration_new'])]

# Group by 'user_id' and count the distinct values in the 'event' column
distinct_event_count = filtered_data.groupby('user_id')['event'].nunique()

# Count the number of users where both 'registration' and 'registration_new' events occurred
count = len(distinct_event_count[distinct_event_count == 2])

print(f"Number of distinct 'user_id's with both 'registration' and 'registration_new' events: {count}")


Number of distinct 'user_id's with both 'registration' and 'registration_new' events: 0


### Now, we will formulate hypotheses  
**H<sub>0</sub> (null hypothesis):** there is no difference in conversion between pages 'registration' and 'registration_new'  
**H<sub>A</sub> (null hypothesis):** there is a difference in conversion between pages 'registration' and 'registration_new'

For testing, we will use a *Chi-squared test* to check independence between registration pages and a *two-sample proportion test* 

In [15]:
from scipy.stats import chi2_contingency

# Create the contingency table
contingency_table = [[24078, 8768], [2540, 1007]]

# Perform the Chi-squared test
chi2, p_value, dof, expected = chi2_contingency(contingency_table)
alpha = 0.05

# Make a decision
if p_value < alpha:
    print("There is a statistically significant difference between the pages.")
else:
    print("There is no statistically significant difference between the pages.")

# Output the test statistics and p-value
print(f"Chi-squared: {chi2}")
print(f"P-value: {p}")


There is a statistically significant difference between the pages.
Chi-squared: 4.600902133327493
P-value: 0.031955136977651455


In [16]:
from statsmodels.stats.proportion import proportions_ztest

# Define the data
count = np.array([8768, 1007])  # Number of logins for control_page and test_page
nobs = np.array([24078, 2540])  # Number of registrations for control_page and test_page

# Perform the two-sample proportion test
z_stat, p_value = proportions_ztest(count, nobs)

# Set the significance level (alpha)
alpha = 0.05

# Make a decision
if p_value < alpha:
    print("There is a statistically significant difference between the pages.")
else:
    print("There is no statistically significant difference between the pages.")
print(f"z-value: {z_stat}")
print(f"P-value: {p_value}")


There is a statistically significant difference between the pages.
z-value: -3.212485940700336
P-value: 0.0013159158900053333


### Conclusion  

We calculated that conversions of 'registration' and 'registration_new' pages are 0.36 and 0.4, respectively, and we see that the latter version of a registration page has better conversion.  
</br>
Hypotheses testing using a *Chi-squared test* and a *two-sample proportion test* showed that difference between conversions is **statistically significant**.   
</br>

Thus, repeating the question posed at the beginning, we can say that **the new registration page is better than the old one**.